runtime/vm/options.c

```diff
@@ -0,0 +1,198 @@
+ *******************************************************************************/
+
+#include <stdio.h>
+#include <assert.h>
+#include "j9protos.h"
+#include "jvminitcommon.h"
+#include "options.h"
+
+
+/*
+ * Example -XX:+PrintFlagsFinal output: (Hotspot calls them "Flags" but OpenJ9 terminology is "options")
+ *
+ * [Global flags]
+ *    size_t MaxHeapSize                              = 4294967296                                {product} {ergonomic}
+ *  uint64_t MaxDirectMemorySize                      = 3758096384                                {product} {default}
+ *
+ * First column is the type of the option
+ * Second column is the name of the option
+ * Third column is the value of the option
+ * Fourth column is a category for the option: product, experimental, etc.
+ * Fifth column describes how the option's value was set: default setting, by command line, or via ergonomics (heuristics)
+ */
+
+
+/* Option types (first ten entries for compatibility) */
+#define TYPE_BOOL 0
+#define TYPE_CCSTR 1
+#define TYPE_CCSTRLIST 2
+#define TYPE_DOUBLE 3
+#define TYPE_INT 4
+#define TYPE_INTX 5
+#define TYPE_SIZET 6
+#define TYPE_UINT 7
+#define TYPE_UINT64T 8
+#define TYPE_UINTX 9
+
+static const char *option_type_string[] = {
+	"bool",
+	"ccstr",
+	"ccstrlist",
+	"double",
+	"int",
+	"intx",
+	"size_t",
+	"uint",
+	"uint64_t",
+	"uintx"
+};
+#define NUM_TYPE_STRINGS	(sizeof(option_type_string) / sizeof(option_type_string[0]))
+
+typedef union J9OptionValue {
+	UDATA v_uintx;
+	U_64 v_uint64;
+	size_t v_sizet;
+	double v_double;
+	char *v_ccstr;
+	char **v_ccstrlist;
+	U_32 v_uint;
+	IDATA v_intx;
+	I_8 v_bool;
+} J9OptionValue;
+
+#define CAT_PRODUCT		0
+#define CAT_ARCHPRODUCT		1
+#define CAT_PDPRODUCT		2
+#define CAT_PRODUCTRW		3
+#define CAT_DIAGNOSTIC		4
+#define CAT_EXPERIMENTAL	5
+#define CAT_MANAGEABLE		6
+
+static const char *option_category_string[] = {
+	"{product}",
+	"{ARCH product}",
+	"{pd product}",
+	"{product rw}",
+	"{diagnostic}",
+	"{experimental}",
+	"{manageable}",
+};
+#define NUM_CATEGORY_STRINGS	(sizeof(option_category_string) / sizeof(option_category_string[0]))
+
+#define SET_DEFAULT	0
+#define SET_COMMANDLINE	1
+#define SET_ERGONOMIC	2
+
+static const char *option_howset_string[] = {
+	"{default}",
+	"{command line}",
+	"{ergonomic}"
+};
+#define NUM_HOWSET_STRINGS	(sizeof(option_howset_string) / sizeof(option_howset_string[0]))
+
+typedef struct J9OptionDesc {
+	const char * name;
+	I_8 type;
+	I_8 category;
+	I_8 howset;
+	J9OptionValue value;
+} J9OptionDesc;
+
+
+/* options_enum gives each option an index starting from zero */
+typedef enum options_enum {
+	NoOption=-1,
+
+	#define OPTION(n,t,cats,how) IDX_ ## n,
+	#include "options.inc"
+	#undef OPTION
+} options_enum;
+
+/* options array lists all options, same order as options_enum */
+static J9OptionDesc options[]={
+	#define OPTION(n,t,cats,how) { n, t, cats, how, { 0 } },
+	#include "options.inc"
```

Comment:
I'd prefer not to introduce options.inc in this PR and just include the 2 values directly here given this is a stepping stone to a larger design.

Given our conversations, I expected that most of this code would be removed and the minimal bits ported to jvminit.c to handle the two options.  And the table-based design would follow after.

Reply:
OK. I'll back everything out and put the three printf's (and code to compute the options) into jvminit.c (so basically back to the absolute minimum hackery to get the three lines of output needed for elasticsearch). I'll push later this afternoon.